backup: TestBackupExportRequestTimeout failed

backup.TestBackupExportRequestTimeout [failed](https://mesolite.cluster.engflow.com/invocations/default/b7a333f3-fce2-490e-8b3e-0940daaeee8d?testReportRun=3&testReportShard=3&testReportAttempt=1#targets-Ly9wa2cvYmFja3VwOmJhY2t1cF90ZXN0) on release-25.1.9-rc @ [2fcef9301ac20e8cb084cf7e2c2c24bede0ee1e7](https://github.com/cockroachdb/cockroach/commits/2fcef9301ac20e8cb084cf7e2c2c24bede0ee1e7):


```
        testing.runTests.func1(0xc0011dcea0)
        	GOROOT/src/testing/testing.go:2161 +0x86
        testing.tRunner(0xc0011dcea0, 0xc003b8f8c8)
        	GOROOT/src/testing/testing.go:1689 +0x21f
        testing.runTests(0xc00154eee8, {0xc001c5b800, 0x7, 0x8}, {0xc0000061c0?, 0xc00154eea0?, 0x10f37bc0?})
        	GOROOT/src/testing/testing.go:2159 +0x8bf
        testing.(*M).Run(0xc001546460)
        	GOROOT/src/testing/testing.go:2027 +0xf18
        github.com/cockroachdb/cockroach/pkg/backup_test.TestMain(0xc001546460)
        	pkg/backup/main_test.go:28 +0x1f1
        main.main()
        	bazel-out/k8-fastbuild/bin/pkg/backup/backup_test_/testmain.go:752 +0x8f4
        goroutine 35 [select]:
        go.opencensus.io/stats/view.(*worker).start(0xc000e16e80)
        	external/io_opencensus_go/stats/view/worker.go:292 +0x1a5
        created by go.opencensus.io/stats/view.init.0 in goroutine 1
        	external/io_opencensus_go/stats/view/worker.go:34 +0xfd
        goroutine 50 [chan receive]:
        github.com/cockroachdb/cockroach/pkg/util/log.flushDaemon()
        	pkg/util/log/log_flush.go:123 +0x6e
        created by github.com/cockroachdb/cockroach/pkg/util/log.init.5 in goroutine 1
        	pkg/util/log/log_flush.go:78 +0x25
        goroutine 51 [chan receive]:
        github.com/cockroachdb/cockroach/pkg/util/log.signalFlusher()
        	pkg/util/log/log_flush.go:147 +0xae
        created by github.com/cockroachdb/cockroach/pkg/util/log.init.5 in goroutine 1
        	pkg/util/log/log_flush.go:79 +0x31
        goroutine 22 [syscall]:
        os/signal.signal_recv()
        	GOROOT/src/runtime/sigqueue.go:152 +0x29
        os/signal.loop()
        	GOROOT/src/os/signal/signal_unix.go:23 +0x1d
        created by os/signal.Notify.func1.1 in goroutine 51
        	GOROOT/src/os/signal/signal.go:151 +0x47
        goroutine 52 [chan receive]:
        github.com/cockroachdb/cockroach/pkg/util/goschedstats.init.0.func1()
        	pkg/util/goschedstats/runnable.go:190 +0x130
        created by github.com/cockroachdb/cockroach/pkg/util/goschedstats.init.0 in goroutine 1
        	pkg/util/goschedstats/runnable.go:182 +0x25
        goroutine 17721 [chan receive]:
        github.com/cockroachdb/pebble/sstable.(*RawColumnWriter).drainWriteQueue(0xc00e4c7008)
        	external/com_github_cockroachdb_pebble/sstable/colblk_writer.go:856 +0x125
        created by github.com/cockroachdb/pebble/sstable.newColumnarWriter in goroutine 17661
        	external/com_github_cockroachdb_pebble/sstable/colblk_writer.go:185 +0x132b: 
        Leaked goroutine: goroutine 17721 [chan receive]:
        github.com/cockroachdb/pebble/sstable.(*RawColumnWriter).drainWriteQueue(0xc00e4c7008)
        	external/com_github_cockroachdb_pebble/sstable/colblk_writer.go:856 +0x125
        created by github.com/cockroachdb/pebble/sstable.newColumnarWriter in goroutine 17661
        	external/com_github_cockroachdb_pebble/sstable/colblk_writer.go:185 +0x132b
--- FAIL: TestBackupExportRequestTimeout (23.15s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=3</code>
 - <code>shard=3</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #150231 backup: TestBackupExportRequestTimeout failed [A-disaster-recovery C-test-failure O-robot T-disaster-recovery branch-release-25.1]
- #146042 backup: TestBackupExportRequestTimeout failed [A-disaster-recovery C-test-failure O-robot P-2 T-disaster-recovery branch-release-25.2]
</p>
</details>
/cc @cockroachdb/disaster-recovery
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestBackupExportRequestTimeout.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-52947

--- a/pkg/backup/backupsink/file_sst_sink.go
+++ b/pkg/backup/backupsink/file_sst_sink.go
@@ -231,11 +231,13 @@ func (s *FileSSTSink) Close() error {
 	if s.cancel != nil {
 		s.cancel()
 	}
+
+	var err error
 	if s.out != nil {
-		return s.out.Close()
+		err = s.out.Close()
 	}
 	s.sst.Close()
-	return nil
+	return err
 }
 
 func (s *FileSSTSink) Flush(ctx context.Context) error {

--- a/pkg/backup/backupsink/file_sst_sink_test.go
+++ b/pkg/backup/backupsink/file_sst_sink_test.go
@@ -441,6 +441,25 @@ func TestFileSSTSinkWrite(t *testing.T) {
 	}
 }
 
+func TestFileSSTSink_CloseLeakRegression(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	// This test is a regression unit test for a resource leak in
+	// FileSSTSink.Close. Previously, if Close was called with an in-complete
+	// flush, it would leak the SST writer.
+
+	ctx := context.Background()
+	st := cluster.MakeTestingClusterSettings()
+	sink, _ := fileSSTSinkTestSetup(t, st, execinfrapb.ElidePrefix_None)
+
+	// Write a span to initialize s.out and s.sst.
+	es := newExportedSpanBuilder("a", "b").withKVs([]kvAndTS{{key: "a", timestamp: 1}}).build()
+	_, err := sink.Write(ctx, es)
+	require.NoError(t, err)
+	_ = sink.Close()
+}
+
 func s2k(s string) roachpb.Key {
 	tbl := 1
 	k := []byte(s)

